lib/workers/repository/update/pr/changelog/source.ts

```diff
@@ -0,0 +1,271 @@
+import URL from 'node:url';
+import is from '@sindresorhus/is';
+import { logger } from '../../../../../logger';
+import { getPkgReleases } from '../../../../../modules/datasource';
+import type { Release } from '../../../../../modules/datasource/types';
+import * as allVersioning from '../../../../../modules/versioning';
+import * as packageCache from '../../../../../util/cache/package';
+import { regEx } from '../../../../../util/regex';
+import { trimSlashes } from '../../../../../util/url';
+import type { BranchUpgradeConfig } from '../../../../types';
+import { slugifyUrl } from './common';
+import { addReleaseNotes } from './release-notes';
+import { getInRangeReleases } from './releases';
+import type {
+  ChangeLogError,
+  ChangeLogRelease,
+  ChangeLogResult,
+} from './types';
+
+export abstract class ChangeLogSource {
+  private platform;
+  private datasource;
+  private cacheNamespace: string;
+
+  constructor(
+    platform: 'bitbucket' | 'github' | 'gitlab',
+    datasource: 'bitbucket-tags' | 'github-tags' | 'gitlab-tags'
+  ) {
+    this.platform = platform;
+    this.datasource = datasource;
+    this.cacheNamespace = `changelog-${platform}-release`;
+  }
+
+  abstract getCompareURL(
+    baseUrl: string,
+    repository: string,
+    prevHead: string,
+    nextHead: string
+  ): string;
+
+  abstract getAPIBaseUrl(config: BranchUpgradeConfig): string;
+
+  async getAllTags(endpoint: string, repository: string): Promise<string[]> {
+    const tags = (
+      await getPkgReleases({
+        datasource: this.datasource,
+        packageName: repository,
+        versioning:
+          'regex:(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?',
```

Comment:
This was one way I found I could get _all_ tags but still use `getPkgReleases`, so to the below existing test and functionality would pass.  

Omitting `versioning` defaulted to semver and truncated any prefixed tags.  

There shouldn't be any need for future overrides

https://github.com/renovatebot/renovate/blob/a68cf49d4311d59186adc2a0c478e92022d956ed/lib/workers/repository/update/pr/changelog/github.spec.ts#L364-L373

Reply:
Not sure if there are limitations for Bitbucket, but in Gitlab the release/tag name can arbitrary. 
Therefore we need to consider supporting e.g. ubuntu versioning in the future. 

Tough I'm fine with approving if we simply extend this for now the current behavior to Bitbucket

Reply:
If you notice, I pushed that logic into the shared `source` file for all three changelog implementations (gitlab, github, bitbucket). 

The way the tag filtering logic works is it still expects it to contain the version _somewhere_ within the tag, so that it can generate the correct compare URL string.

I don't believe the current behavior would support these tag types